File: src/ServerScriptService/dataStore.server.lua
local dataStore = game:GetService("DataStoreService"):GetDataStore("SaveData")

game.Players.PlayerAdded:Connect(function(player)
	-- Variables
	wait()
	local moneyValue = player.leaderstats.Money
	local tierValue = player.leaderstats.Tier	
	
	-- Save with the id of the player to recognise who the data belongs to
	local attempt = 0
	local success = false
	local getSaved
	-- tente d'avoir les sauvegarde déjà effectué
	repeat
		success, getSaved = pcall(function()
			return dataStore:GetAsync(player.UserId)
		end)
		attempt += 1
	until success or attempt == 5
	
	if getSaved then
		moneyValue.Value = getSaved[1]
		tierValue.Value = getSaved[2]
	else
		moneyValue.Value = 1000
		tierValue.Value = 0
	end
end)

game.Players.PlayerRemoving:Connect(function(player)
	local attempt = 0
	repeat
		local success, message = pcall(function()
			return dataStore:SetAsync(player.UserId, {player.leaderstats.Money.Value, player.leaderstats.Tier.Value})
		end)
		attempt += 1
	until success or attempt >= 5 
end)

File: src/ServerScriptService/dataStore.server.ts
import { Players, DataStoreService } from "@rbxts/services";

const dataStore = DataStoreService.GetDataStore("SaveData");

Players.PlayerAdded.Connect((player) => {
    // Variables
    const moneyValue = player.FindFirstChild("leaderstats")?.FindFirstChild("Money") as IntValue;
    const tierValue = player.FindFirstChild("leaderstats")?.FindFirstChild("Tier") as StringValue;

    if (!moneyValue || !tierValue) return;

    // Save with the id of the player to recognize who the data belongs to
    let attempt = 0;
    let success = false;
    let getSaved: unknown;

    // Try to get already saved data
    do {
        [success, getSaved] = pcall(() => {
            return dataStore.GetAsync(tostring(player.UserId));
        });
        attempt += 1;
    } while (!success && attempt < 5);

    if (getSaved) {
        moneyValue.Value = (getSaved as Array<number>)[0];
        tierValue.Value = (getSaved as Array<string>)[1];
    } else {
        moneyValue.Value = 1000;
        tierValue.Value = "1";
    }
});

Players.PlayerRemoving.Connect((player) => {
    const moneyValue = player.FindFirstChild("leaderstats")?.FindFirstChild("Money") as IntValue;
    const tierValue = player.FindFirstChild("leaderstats")?.FindFirstChild("Tier") as StringValue;

    if (!moneyValue || !tierValue) return;

    let attempt = 0;
    let success = false;

    do {
        [success] = pcall(() => {
            return dataStore.SetAsync(tostring(player.UserId), [moneyValue.Value, tierValue.Value]);
        });
        attempt += 1;
    } while (!success && attempt < 5);
});
